File: programming-language/Test/numerical_methods/question11.R
# Função que define o sistema de equações e sua matriz jacobiana
equations <- function(vars) {
  x <- vars[1]
  y <- vars[2]
  
  f <- c(
    x^3 + y - 1,
    y^3 - x + 1
  )
  
  J <- matrix(0, nrow = 2, ncol = 2)
  J[1, 1] <- 3 * x^2
  J[1, 2] <- 1
  J[2, 1] <- -1
  J[2, 2] <- 3 * y^2
  
  list("f" = f, "J" = J)
}

# Função que implementa o método de Newton
newton_method <- function(equations, x0, epsilon = 0.0001, max_iter = 1000) {
  x <- x0
  for (i in 1:max_iter) {
    result <- equations(x)
    f <- result$f
    J <- result$J
    delta <- solve(J, -f)
    x <- x + delta
    if (max(abs(delta)) < epsilon) {
      break
    }
  }
  return(x)
}

# Valor inicial
x0 <- c(1, 1)

# Chamar o método de Newton
solution <- newton_method(equations, x0)

# Exibir a solução
print(solution)

File: programming-language/Test/numerical_methods/question11.py
import numpy as np


# Funções que representam as equações do sistema
def f(x, y):
    return x ** 3 + y - 1


def g(x, y):
    return y ** 3 - x + 1


# Derivadas parciais das funções em relação a x e y
def df_dx(x, y):
    return 3 * x ** 2


def df_dy(x, y):
    return 1


def dg_dx(x, y):
    return -1


def dg_dy(x, y):
    return 3 * y ** 2


# Implementação do método de Newton
def newton_method(x0, y0, epsilon=0.0001, max_iterations=100):
    x = x0
    y = y0
    iterations = 0

    while True:
        fx = f(x, y)
        gy = g(x, y)

        J = np.array([[df_dx(x, y), df_dy(x, y)],
                      [dg_dx(x, y), dg_dy(x, y)]])

        J_inv = np.linalg.inv(J)

        delta = np.dot(J_inv, np.array([-fx, -gy]))

        x += delta[0]
        y += delta[1]

        iterations += 1

        if np.linalg.norm(delta) < epsilon or iterations >= max_iterations:
            break

    return x, y


# Escolha dos valores iniciais
x0 = 1
y0 = 1

# Resolução do sistema de equações
x_hat, y_hat = newton_method(x0, y0)

print("Raízes encontradas:")
print("x̂ = {:.3f}".format(x_hat))
print("ŷ = {:.3f}".format(y_hat))
